### PR TITLE
Improve partitioned output operator tests

### PR DESCRIPTION
**Pull request by lukasz-stec**
Added a set of query tests in TestRepartitionQueries to verify PartitioneOutOperator.
TestPartitionedOutputOperator was refactored and improved with more cases and more detailed assertions about how the operator partitions the input positions.

### CODE REVIEW

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java [hunk 1]
```diff
@@ -158,6 +162,21 @@ public static Block createRandomBlockForType(Type type, int positionCount, float
         if (type == VARCHAR) {
             return createRandomStringBlock(positionCount, nullRate, MAX_STRING_SIZE);
         }
+        if (type instanceof CharType) {
```
Comment (joshthoward) [inline core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:165]: Why is there no `CharType.CHAR` field like with the other types?

Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:165]: `CharType` is parametrized (`length` param) + it does not have `UNBOUNDED_LENGTH` like `VarcharType`

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java [hunk 2]
```diff
@@ -435,6 +473,11 @@ public static Block createLongDecimalsBlock(Iterable<String> values)
         return builder.build();
     }
 
+    public static Block createCharsBlock(CharType charType, List<String> values)
+    {
+        return createBlock(charType, charType::writeString, values);
```
Comment (skrzypo987) [inline core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:478]: For chars you should also right trim the ' ' characters (if any))

File: core/trino-main/src/test/java/io/trino/operator/output/TestPartitionedOutputOperator.java [hunk 3]
```diff
@@ -13,248 +13,693 @@
...
+        processPages(partitionedOutputOperator, page);
+
+        List<Object> partition0 = readLongs(outputBuffer.getEnqueuedDeserialized(0), 0);
+        assertThat(partition0).containsExactlyInAnyOrder(4L, 6L);
+        List<Object> partition1 = readLongs(outputBuffer.getEnqueuedDeserialized(1), 0);
+        assertThat(partition1).containsExactlyInAnyOrder(5L, 7L);
+    }
+
+    @DataProvider(name = "types")
+    public static Object[][] types()
```
Comment (skrzypo987) [inline core/trino-main/src/test/java/io/trino/operator/output/TestPartitionedOutputOperator.java:364]: Move that below `testOutputForSimplePageWithType`

Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/operator/output/TestPartitionedOutputOperator.java:379]: done

File: testing/trino-tests/src/test/java/io/trino/tests/TestRepartitionQueries.java [hunk 4]
```diff
@@ -0,0 +1,155 @@
...
+    }
+
+    private void testRepartitioning(String columnExpression)
+    {
+        testRepartitioning(columnExpression, "column_under_test", columnExpression);
+    }
+
+    private void testRepartitioning(String columnExpression, String columnProjection, String actualColumnExpression)
+    {
+        assertQuery(format("WITH custkey_ex AS (" +
```
Comment (sopel39) [inline testing/trino-tests/src/test/java/io/trino/tests/TestRepartitionQueries.java:140]: add some comment how does this assertion work

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java [hunk 5]
```diff
@@ -780,6 +795,27 @@ public static RunLengthEncodedBlock createRLEBlock(long value, int positionCount
         return new RunLengthEncodedBlock(blockBuilder.build(), positionCount);
     }
 
+    private interface ValueWriter<T>
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:798]: move it below usage (`createBlock`)

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java [hunk 6]
```diff
@@ -660,18 +681,12 @@ public static Block createDoublesBlock(Double... values)
...
-            }
-            else {
-                DOUBLE.writeDouble(builder, value);
-            }
-        }
+        return createDoublesBlock(DOUBLE, values);
+    }
 
-        return builder.build();
+    public static Block createDoublesBlock(Type type, Iterable<Double> values)
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/block/BlockAssertions.java:687]: type is not needed here, is it? Only `DOUBLE` uses `double` primitive

File: core/trino-main/src/test/java/io/trino/operator/output/TestPartitionedOutputOperator.java [hunk 7]
```diff
@@ -13,248 +13,693 @@
...
-    protected TestPartitionedOutputOperator(OperatorFactories operatorFactories)
+    protected TestPartitionedOutputOperator(Session testSession, OperatorFactories operatorFactories)
     {
+        this.testSession = testSession;
         this.operatorFactories = operatorFactories;
     }
 
     @BeforeClass
-    public void setUp()
+    public void setUpClass()
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/output/TestPartitionedOutputOperator.java:131]: undo

File: core/trino-main/src/test/java/io/trino/operator/output/TestPartitionedOutputOperator.java [hunk 8]
```diff
@@ -13,248 +13,693 @@
...
     {
         executor = newCachedThreadPool(daemonThreadsNamed(getClass().getSimpleName() + "-executor-%s"));
         scheduledExecutor = newScheduledThreadPool(1, daemonThreadsNamed(getClass().getSimpleName() + "-scheduledExecutor-%s"));
     }
 
     @AfterClass(alwaysRun = true)
-    public void tearDown()
+    public void tearDownClass()
```
Comment (sopel39) [inline core/trino-main/src/test/java/io/trino/operator/output/TestPartitionedOutputOperator.java:138]: undo

**Review by sopel39 — APPROVED** [review]
lgtm % comments

**Comment by sopel39** [comment]
mind automation

File: core/trino-main/src/test/java/io/trino/block/BlockAssertions.java (hunk 6 above)
Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/block/BlockAssertions.java]: yup, it's not needed. I also remove unnecessary type from other methods

File: core/trino-main/src/test/java/io/trino/operator/output/TestPartitionedOutputOperator.java (hunk 7 above)
Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/operator/output/TestPartitionedOutputOperator.java:131]: I added `@BeforeMethod setUp` and it clashes with this name. I think `setUpClass` is a better name for  `@BeforeClass`

File: core/trino-main/src/test/java/io/trino/operator/output/TestPartitionedOutputOperator.java (hunk 8 above)
Comment (lukasz-stec) [reply core/trino-main/src/test/java/io/trino/operator/output/TestPartitionedOutputOperator.java:138]: its for consistency with `setUpClass`

**Review by lukasz-stec — COMMENTED** [review]
comments addressed. new commits squashed into old ones

**Comment by lukasz-stec** [comment]
checkstyle fix

**Review by sopel39 — APPROVED** [review]

**Pull request merged**